File: etc/gen.py

from ast import literal_eval as lit
import sys
import re
import dataclasses
from typing import List, Tuple

CONST_RE = re.compile(
    r"const\s+(?P<name>\w+)\s*:"
    + r"\s*(?P<ty>\w+)\s*="
    + r"\s*(?P<value>\w+)*")
STRUCT_RE = re.compile(r"struct\s+(?P<name>\w+)\s*\[(?P<body>.*?)\]", re.DOTALL)
_TY = r""
TY_PAT = r"(?P<refs>&*)\s*(?P<ty>\w+)"
FIELD_RE = re.compile(r"(?P<name>\w+)\s*:\s*" + TY_PAT, re.MULTILINE)

assert CONST_RE.search("const foo : U2 = 0x7")

assert STRUCT_RE.search("struct Foo [ a: U2;  b: Arr[4 U2]]").group('name', 'body') == (
    'Foo', ' a: U2;  b: Arr[4 U2')

NATIVE_TYS = {"U1", "U2", "U4", "S", "I1", "I2", "I4"}

@dataclasses.dataclass
class Const:
  name: str; ty: str; value: str

  def isNative(self):
    return self.ty in NATIVE_TYS

@dataclasses.dataclass
class TyI:
  refs: int; ty: str

@dataclasses.dataclass
class Field:
  name: str
  refs: int
  ty: str

@dataclasses.dataclass
class Struct:
  name: str;
  fields: List[Field]

  def clang(self, f):
    f.write('typedef struct {\n')
    for field in self.fields:
      refs = '*' * field.refs
      f.write(f'{field.ty}{refs} {field.name};\n')
    f.write(f'}} {self.name};\n')

def parseTyI(s: str):
  refs, ty = TYI_RE.match(s).group("refs", "ty")
  refs = int(refs) if refs else 0
  return TyI(refs, ty)

@dataclasses.dataclass
class Parsed:
  consts: List[Tuple[str, int]]
  structs: List[Struct]

def parseFngi(path):
  with open(path) as f: text = f.read()
  parsed = [Const(*m.group("name", "ty", "value")) for m in CONST_RE.finditer(text)]
  consts = [(c.name, lit(c.value)) for c in parsed if c.isNative()]

  structs = []
  for sm  in STRUCT_RE.finditer(text):
    fields = []
    for m in FIELD_RE.finditer(sm.group('body')):
      name, refs, ty = m.group("name", "refs", "ty")
      fields.append(Field(name, len(refs) if refs else 0, ty))
    structs.append(Struct(sm.group('name'), fields))

  return Parsed(consts = consts, structs=structs)

def writeConsts(f, path, consts):
  f.write(f"// GENERATED BY etc/gen.py FROM {path}\n// DO NOT EDIT MANUALLY\n\n")
  for name, value in consts:
    f.write(f"#define {name:<16} {hex(value)}\n")

sporPath = 'src/spor.fn'; spor = parseFngi(sporPath)
with open("gen/spor.h", "w") as f: writeConsts(f, sporPath, spor.consts)

sz = {name: value for (name, value) in spor.consts if name.startswith('SZ')}
instrs = [(instr, name) for (name, instr) in spor.consts if name not in sz]
instrs.sort()

sporConsts = dict(spor.consts)
SLIT = sporConsts['SLIT']

SZ_MASK = sporConsts['SZ_MASK']
SZ1 = sz['SZ1']
SZ2 = sz['SZ2']
SZ4 = sz['SZ4']

unsized = {'LCL', 'XL', 'XW', 'XLL', 'XRL', 'SLIT'}

def writeCase(f, name, ret=None):
  ret = ret or name
  f.write(f'    case {name:<16}: return Slc_ntLit("{ret}");\n')

def withSize(f, name, instr):
  for sz in ('1', '2', '4'):
    writeCase(f, f'{name} + SZ{sz}', name + sz)

def slitCases(f):
  for v in range(0, 0x30):
    writeCase(f, f'SLIT + 0x{v:X}', f'{{0x{v:02X}}}')

with open('gen/name.c', 'w') as f:
  f.write('/* Custom generated by etc/gen.py */\n\n')
  f.write('#include "civ.h"\n')
  f.write('#include "spor.h"\n\n')
  f.write('/*extern*/ U1* unknownInstr = "UNKNOWN";\n\n')
  f.write('Slc instrName(U1 instr) {\n')
  f.write('  switch(instr) {\n')
  for instr, name in instrs:
    if name == 'SLIT': slitCases(f)
    elif instr < 0x40 or name in unsized: writeCase(f, name)
    else:                                 withSize(f, name, instr)

  f.write('  }\n')
  f.write('  return (Slc) {.dat = unknownInstr, .len = 7};\n')
  f.write('}\n')



from pprint import pprint
datPath = 'src/dat.fn'; dat = parseFngi(datPath)
pprint(dat.structs)

compPath = 'src/comp.fn'; comp = parseFngi(compPath)
with open("gen/comp.h", "w") as f:
  writeConsts(f, compPath, comp.consts)

File: etc/gen.lua
require('gciv')

local CONST_PAT = (
  'const%s+([_%w]+)%s*:'
  .. '%s*([_%w]+)%s*='
  .. '%s*([_%w]+)'
)
local STRUCT_PAT = 'struct%s+(%w+)%s*(%b[])'
local TY_PAT = '(&*)%s*([_%w]+)'
local FIELD_PAT = '([_%w]+)%s*:%s*' .. TY_PAT

local NATIVE_TYS = Set{"U1", "U2", "U4", "S", "I1", "I2", "I4"}

local UNSIZED = Set{'MMV', 'LCL', 'XL', 'XW', 'XLL', 'XRL', 'SLIT'}

local RENAME = {
    Arena  = "SpArena",
    Reader = "SpReader",
    Writer = "SpWriter",
    Fmt    = "SpFmt",
    Logger = "SpLogger",
}

local tyDictDefined = false

local FConst = struct('FConst', {'name', 'ty', 'value'})
-- Const: {name, ty, value}
local function parseConsts(s, to)
  if not to then to = List{} end
  for name, ty, value in string.gmatch(s, CONST_PAT) do
    to:add(FConst{name=name, ty=ty, value=value})
  end
  return to
end

local FField = struct('', {
  {'name', String}, {'refs', Int}, {'ty', String}
})
local FStruct = struct('FStruct', {
  {'name', String}, {'fields', List}
})

local allStructs = Map{}
local function parseStructs(s, to)
  if not to then to = List{} end
  for sname, body in string.gmatch(s, STRUCT_PAT) do
    local fields = List{}
    for name, refs, ty in string.gmatch(body, FIELD_PAT) do
      if name == 'parent' then fields:extend(allStructs[ty].fields)
      else
        fields:add(FField{
          name=name, refs=string.len(refs), ty=ty,
        })
      end
    end
    local st = FStruct{name=sname, fields=fields}
    allStructs[sname] = st; to:add(st)
  end
  return to
end

local function parseFngi(path, to)
  local text = readAll(path)
  if not to then
    to = Map{consts=List{}, structs=List{}}
  end
  parseConsts(text,  to.consts)
  parseStructs(text, to.structs)
  return to
end


local spor = parseFngi('src/spor.fn')
local dat = parseFngi('src/dat.fn')

local datStructs = copy(allStructs)
local comp = parseFngi('src/comp.fn')
local compStructs = allStructs:diff(datStructs)

local function writeConsts(f, path, consts)
  f:write'// DO NOT EDIT MANUALLY! GENERATED BY etc/gen.lua\n'
  f:write('// See docs at: ' .. path .. '\n')
  f:write'#include "civ.h"\n\n'
  for _, c in ipairs(consts) do
    f:write('#define ' .. c.name .. '  ' .. c.value .. '\n')
  end
  return b
end

RENAME = {
    Arena=  "SpArena",
    Reader= "SpReader",
    Writer= "SpWriter",
    Fmt=    "SpFmt",
    Logger= "SpLogger",
}


local function writeStruct(f, s, needDeclare)
end

local function writeStructs(f, structs, needDeclare)
  local needDeclare = needDeclare or {}
  for _, s in ipairs(structs) do
    local sname = RENAME[s.name] or s.name
    f:write(concat{'\ntypedef struct _', sname, ' {\n'})
    for _, field in ipairs(s.fields) do
      local ty_ = RENAME[field.ty] or field.ty
      if ty_ == 'Self' then ty_ = 'struct _' .. sname end
      if needDeclare[ty_] then ty_ = 'struct _' .. ty_ end
      f:write(concat{'  ', ty_}); for _=1, field.refs do f:write'*' end
      f:write(concat{' ', field.name, ';\n'})
    end
    f:write(concat{'} ', sname, ';\n'})
    needDeclare[sname] = nil
  end
end

local COMP_H = [=====[

// -- this part is actually hand written in gen.lua --
struct _TyDict;
struct _TyFn;

typedef struct {
  struct _TyFn*  drop;     // this:&This -> ()
  struct _TyFn*  alloc;    // this:&This sz:S alignment:U2 -> Ref
  struct _TyFn*  free;     // this:&This dat:Ref sz:S alignment:U2 -> ()
  struct _TyFn*  maxAlloc; // this:&This -> S
} MSpArena;
typedef struct { void* d; MSpArena* m; } SpArena;

typedef struct {
  struct _TyFn*  read;   // this:&This -> ()
  struct _TyFn*  asBase; // this:&This -> &BaseFile
} MSpReader;
typedef struct { void* d; MSpReader* m; } SpReader;

typedef struct {
  struct _TyFn*  asBase; // this:&This -> &BaseFile
  struct _TyFn*  write;  // this:&This -> ()
} MSpWriter;
typedef struct { void* d; MSpWriter* m; } SpWriter;

typedef struct {
  MSpWriter w;
  struct _TyFn*  state;  // this:&This -> &FmtState
} MSpFmt;
typedef struct { void* d; MSpFmt* m; } SpFmt;

typedef struct {
  MSpFmt         fmt;
  struct _TyFn*  logConfig;  // this:&This -> &LogConfig

  struct _TyFn*  start;      // this:&This U1  -> U1
  struct _TyFn*  add;        // this:&This Slc -> ()
  struct _TyFn*  end;        // this:&This     -> ()
} MSpLogger;
typedef struct { void* d; MSpLogger* m; } SpLogger;

// -- everything after this is truly generated --
]=====]

local NAME_C = [===[
/* Custom generated by etc/gen.lua */

#include "civ.h"
#include "spor.h"

/*extern*/ U1* unknownInstr = "UNKNOWN";

Slc instrName(U1 instr) {
  switch(instr) {
]===]


local function writeCase(f, name, ret)
  ret = ret or name
  f:write(concat{'    case ', name, ': return Slc_ntLit("', ret, '");\n'})
end

local function withSize(f, name)
  for _, sz in ipairs({'1', '2', '4'}) do
    writeCase(f, string.format('%s + SZ%s', name, sz), name .. sz)
  end
end

local function slitCases(f)
  for v=0,0x29 do
    writeCase(f, string.format('SLIT + 0x%X', v), string.format('0x%X', v))
  end
end

local sz, instrs = Map{}, List{}
for _, c in pairs(spor.consts) do
  if string.match(c.name, '^SZ') then sz[c.name] = c.value
  else instrs:add({instr=tonumber(c.value), name=c.name}) end
end

local function writeCLang(f, path, mod)
  writeConsts(f,  path, mod.consts)
  writeStructs(f, mod.structs)
  return f
end

local function main()
  print("## Generating C using lua")
  local f = io.open('gen/name.c', 'w')
  f:write(NAME_C)
  for _, i in ipairs(instrs) do
    if i.name == 'SLIT'                        then slitCases(f)
    elseif (i.instr < 0x40) or UNSIZED[i.name] then writeCase(f, i.name)
    else                                withSize(f, i.name, i.instr)
    end
  end
  f:write'  }\n'
  f:write'  return (Slc) {.dat = unknownInstr, .len = 7};\n'
  f:write'}\n'
  f:close()

  local sporPath, compPath = 'gen/spor.h', 'gen/comp.h'
  writeCLang(io.open(sporPath, 'w'), sporPath, spor):close()
  f = io.open(compPath, 'w')
  writeConsts(f,  compPath, comp.consts)
  f:write(COMP_H)

  writeStructs(f, comp.structs, {TyDict=1, TyFn=1})
  f:close()
end


if not TESTING then main() end
return {
  parseConsts = parseConsts,
  parseStructs = parseStructs,
  STRUCT_PAT = STRUCT_PAT,
  FIELD_PAT = FIELD_PAT,
  FConst = FConst, FField = FField, FStruct = FStruct,
  spor = spor, dat = dat, comp=comp,
  datStructs = datStructs, compStructs = compStructs,

  genConsts = genConsts,
}
